pkg/providers/gcp/gcp.go

```diff
@@ -0,0 +1,156 @@
+package gcp
+
+import (
+	"context"
+	"fmt"
+
+	monitoring "cloud.google.com/go/monitoring/apiv3/v2"
+	monitoringpb "cloud.google.com/go/monitoring/apiv3/v2/monitoringpb"
+	"google.golang.org/api/iterator"
+)
+
+var (
+	/*
+	* An MQL query that will return data from Google Cloud with the
+	* - Instance Name
+	* - Region
+	* - Reserved CPU Cores
+	 */
+	CPUCountQuery = `
+		fetch gce_instance::compute.googleapis.com/instance/cpu/reserved_cores
+		| filter project_id = '%s' 
+		| group_by [metric.instance_name, metadata.system.region ], max(value.reserved_cores)
+		| window(%s)
+		| within %s
+	`
+	/*
+	* An MQL query that will return data from Google Cloud with the
+	* - Instance Name
+	* - Region
+	* - CPU Utilization (as a percentage)
+	 */
+	CPUUtilizationQuery = `
+		fetch gce_instance::compute.googleapis.com/instance/cpu/utilization 
+		| filter project_id = '%s' 
+		| group_by [instance_name, metadata.system.region ], max(value.utilization)
+		| window(%s)
+		| within %s
+	`
+)
+
+// GCP is the structure used as the provider for Google Cloud Platform
+type GCP struct {
+	client    *monitoring.QueryClient
+	projectID string
+}
+
+// TODO: Placeholder, make use of global types when they are ready
+type CPUUtilizationMetric struct {
+	InstanceName string
+	Utilization  float64
+	Region       string
+}
+
+// TODO: Placeholder, make use of global types when they are ready
+type CPUCountMetric struct {
+	InstanceName string
+	Count        float64
+	Region       string
+}
+
+type options func(*GCP)
+
+// WithProjectID is an option used to pass a projectID to the provider
+func WithProjectID(projectID string) options {
+	return func(g *GCP) {
+		g.projectID = projectID
+	}
+}
+
+// New returns a new instance of the GCP provider as well as a function to
+// cleanup connections once done
+func New(ctx context.Context, opts ...options) (g *GCP, teardown func(), err error) {
+	// set any defaults here
+	g = &GCP{}
+
+	// overwrite any options
+	for _, opt := range opts {
+		opt(g)
+	}
+
+	// This allows overwriting the default client
```

Comment:
Wouldn't the default be this functionality, and then the opts would overwrite it? It seems like you're setting it to nil, then having the options overwrite it (if passed), then checking it again, and setting a new default if not set.Or am I missing something?

Reply:
The issue with this client is that it tries authentication on initialization meaning you can't init it without credentials which breaks testing, not sure how to get around it 😅

Reply:
But I agree, that was my first implementation